api/spec/src/subscription.tsp

```diff
@@ -0,0 +1,266 @@
+
+  /**
+   * Cancels the subscription.
+   */
+  @post
+  @operationId("cancelSubscription")
+  @route("/{subscriptionId}/cancel")
+  cancel(
+    @path subscriptionId: ULID,
+    @body body: {
+      effective: "IMMEDIATE" | DateTime;
+    },
+  ): Subscription | OpenMeter.ConflictError | OpenMeter.NotFoundError | OpenMeter.CommonErrors;
+
+  /**
+   * Cancels the subscription.
+   */
+  @post
+  @operationId("cancelScheduledCancelation")
+  @route("/{subscriptionId}/cancel-scheduled-cancelation")
+  cancelScheduledCancelation(@path subscriptionId: ULID):
+    | Subscription
+    | OpenMeter.ConflictError
+    | OpenMeter.NotFoundError
+    | OpenMeter.CommonErrors;
+
+  @get
+  @operationId("getSubscriptionItemsAt")
+  @route("/{subscriptionId}/items")
+  getItems(
+    @path subscriptionId: ULID,
+    @query at: DateTime,
+  ): SubscriptionItem[] | OpenMeter.NotFoundError | OpenMeter.CommonErrors;
+
+  @get
+  @operationId("getPhases")
+  @summary("Get the phases of a subscription")
+  @route("/{subscriptionId}/phases")
+  getPhases(@path subscriptionId: ULID): SubscriptionPhase[] | OpenMeter.NotFoundError | OpenMeter.CommonErrors;
+
+  @delete
+  @operationId("deleteSubscriptionPhase")
+  @route("/{subscriptionId}/phases/{phaseKey}")
+  deletePhase(@path subscriptionId: ULID, @path phaseKey: Key): {
+    @statusCode _: 204;
+  } | OpenMeter.NotFoundError | OpenMeter.CommonErrors;
+
+  @post
+  @operationId("addSubscriptionPhase")
+  @route("/{subscriptionId}/phases")
+  addPhase(
+    @path subscriptionId: ULID,
+    @body phase: {
+      key: Key;
+      rateCards: RateCard[]; // TODO: we need common ground on this interface
+      activeFrom: DateTime;
+    },
+  ): SubscriptionPhase | OpenMeter.ConflictError | OpenMeter.CommonErrors;
+
+  /**
+   * Extend has to delay both the current and all subsequent phases by the provided duration.
+   */
+  @post
+  @operationId("extendSubscriptionPhase")
+  @route("/{subscriptionId}/phases/{phaseKey}/extend")
+  extendPhase(
+    @path subscriptionId: ULID,
+    @path phaseKey: Key,
+    @body body: {
+      @encode(DurationKnownEncoding.ISO8601)
+      extendBy: duration;
+    },
+  ): SubscriptionPhase[] | OpenMeter.ConflictError | OpenMeter.CommonErrors;
+
+  @post
+  @operationId("addSubscriptionPhaseRateCard")
+  @route("/{subscriptionId}/phases/{phaseKey}/items")
+  addPhaseRateCard(
+    @path subscriptionId: ULID,
+    @path phaseKey: Key,
+    @body rateCard: RateCard,
+  ): SubscriptionItem | OpenMeter.ConflictError | OpenMeter.CommonErrors;
+
+  @delete
+  @operationId("deleteSubscriptionPhaseItem")
+  @route("/{subscriptionId}/phases/{phaseKey}/items/{itemKey}")
+  deletePhaseItem(@path subscriptionId: ULID, @path phaseKey: Key, @path itemKey: Key): {
+    @statusCode _: 204;
+  } | OpenMeter.NotFoundError | OpenMeter.CommonErrors;
+}
+
+/**
+ * Subscription is an exact subscription instance.
+ */
+@friendlyName("Subscription")
+model Subscription extends global.Resource {
```

Comment:
Afaik this should be used with spreading. Am I right @tothandras ?

Reply:
Yes